### PR TITLE
Migrate to pytorch-transformers 1.2

### PR DESCRIPTION
**Pull request by sshleifer**
### S3 Changes
In order to run `convai_evaluation.py` or `interact.py` on the existing checkpoint, we need a few changes:

1. adding vocab_size = 40483 to config.json
 - 	I tried to avoid changing this, but the embeddings won't load at the wrong size (40478), and sending `n_special` to the model doesn't do anything, so this seemed like the easiest route. I might be missing a cleaner way.
2. renaming some parameters in `pytorch_model.bin`
3. replace `special_tokens.txt` with `added_tokens.json`, containing the following

```
{"<bos>": 40478, "<eos>": 40479, "<speaker1>": 40480, "<speaker2>": 40481, "<pad>": 40482}
```
to ensure alignment.

Here is the updated [cache](https://www.dropbox.com/s/bt6n0kyqsrnyx3e/gpt_personachat_cache.tar.gz?dl=0)

### Code changes
* `train.py` always tries to add special tokens, then resizes embeddings if any special tokens were added, with the following logic:

```
orig_num_tokens = len(tokenizer.encoder)
num_added_tokens = tokenizer.add_special_tokens(SPECIAL_TOKENS_DICT)
if num_added_tokens > 0:
    model.resize_token_embeddings(new_num_tokens=orig_num_tokens + num_added_tokens)
```
* call to TensorboardLogger.writer.log_dir used to raise AttributeError, it's called `writer.logdir` on that repo's master now. requirements.txt reflects this.

* Switched `OpenAIAdam()` -> `AdamW(correct_bias=True)`. Did not change `PiecewiseLinearScheduler`. My reasoning was that   (a) old code warned that "t_total value of -1 results in schedule not being applied", from which I infer that PiecewiseLinear (from ignite) was doing all the scheduling (not the `schedule='warmup_linear'` kwarg in `OpenAIAdam`. Learning rates in tensorboards from before and after the change support this theory; they are identical and go straight down without warmup.

* I added one unittest in `test_tokenizers.py` to make sure I wasn't messing up the tokenizers.

* `convai_evaluation.py` supports GPT2 model checkpoints

### Sanity Check
TLDR: metrics and generations are similar but not identical after training for 1 epoch  (2 V100s,  FP16='O1').

**Before Change:**
- Validation: {'accuracy': 0.7413,
 'average_accuracy': 0.7368,
 'average_nll': 2.6829,
 'average_ppl': 14.6283,
 'nll': 2.6755}
- convai_evaluation: {'exs': 7801, 'hits@1': **0.737**, 'hits@5': 0.955, 'hits@10': 0.99, 'hits@100': 1.0}
- 44 mins

**After Change:**

- Validation: {'accuracy': 0.7657,
 'average_accuracy': 0.7582,
 'average_nll': 2.63606,
 'average_ppl': 13.9582,
 'nll': 2.6283}

- convai_evaluation: {'exs': 7801, 'hits@1': **0.758**, 'hits@5': 0.957, 'hits@10': 0.992, 'hits@100': 1.0}
- 42 mins

I don't have the full output, but hits@1 for the cached/competition model stays above 79 after the change.

### Unfinished
- update S3 and change `S3_PATH`
- `interact.py` for GPT2 (this is also broken before the change)

Feedback much appreciated!

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -141,15 +143,20 @@ def train():
         args.device = torch.device("cuda", args.local_rank)
         torch.distributed.init_process_group(backend='nccl', init_method='env://')
 
-    logger.info("Prepare tokenizer, pretrained model and optimizer - add special tokens for fine-tuning")
```
Comment (sshleifer) [inline train.py]: There is a separate logging message now about special tokens.

**Comment by kiddyboots216** [comment]
Does this make train.py fully compatible with gpt2?

**Comment by sshleifer** [comment]
@kiddyboots216  I believe so. I tested for 1 epoch on a small dataset and nothing broke. `interact.py` still doesn't work, however.

**Comment by kiddyboots216** [comment]
@sshleifer can you give the command needed to reproduce? I get:
```
AttributeError: 'GPT2DoubleHeadsModel' object has no attribute 'set_num_special_tokens'
```
When running:
```
python -m torch.distributed.launch --nproc_per_node=8 ./train.py --gradient_accumulation_steps=4 --lm_coef=2.0 --max_history=2 --n_epochs=1 --num_candidates=4 --personality_permutations=2 --train_batch_size=2 --valid_batch_size=2 --model_checkpoint gpt2
```

**Comment by sshleifer** [comment]
It sounds to me like your command is fine but you are not on this branch.

**Comment by kiddyboots216** [comment]
Of course. Thanks for the clarification.

**Comment by kiddyboots216** [comment]
@sshleifer what command are you running to see evaluation results? I believe `convai_evaluation.py` needs to be changed in order to accomodate GPT2; I imported the GPT2Tokenizer, GPT2DoubleHeadsModel, GPT2LMHeadModel and then create them instead of the OpenAIGPT versions. This gives terrible results even when the GPT2 pretrained seems to get decent validation performance.

**Comment by sshleifer** [comment]
I haven't fixed evaluation with GPT2 in this PR, sorry.

**Comment by kiddyboots216** [comment]
Ok. I have it *technically* working but the results are poor. Also, a note -in adjusting evaluation to work with finetuned GPT models, I think you may have broken evaluation for the base model (if you pass in "openai-gpt" for model_checkpoint there's an error bc the vocab size is incorrect)

**Comment by sshleifer** [comment]
On branch `migrate-pytorch-transformers`

```bash
cd ParlAI
python ../convai_evaluation.py --model_checkpoint openai-gpt
``` 
works for me. Is that what you are running?

**Comment by kiddyboots216** [comment]
Yes, and my results from running that are nowhere near what you are reporting. The hits@1 is <10%.

**Comment by sshleifer** [comment]
I get roughly that (bad) score on both branches, so it doesn't seem like a bug was introduced. Rather, the model requires finetuning to perform well. The performance I report is for the **finetuned** model. I updated the PR description to make that clearer.

**Comment by fedpbt** [comment]
I see -I thought passing in openai-gpt would run evaluation with the competition model. After fine-tuning GPT2 for 1 epoch I get slightly better validation stats in all categories than what you posted, but simply changing the Tokenizer and DoubleHeadsModel from OpenAIGPT -> GPT2 allows the model to run, but I get very poor evaluation results (8% hits @ 1). I'd like to fix this myself but not sure where to start, do you have any ideas?

**Comment by kiddyboots216** [comment]
see -I thought passing in openai-gpt would run evaluation with the competition model. After fine-tuning GPT2 for 1 epoch I get slightly better validation stats in all categories than what you posted, but simply changing the Tokenizer and DoubleHeadsModel from OpenAIGPT -> GPT2 allows the model to run, but I get very poor evaluation results (8% hits @ 1). I'd like to fix this myself but not sure where to start, do you have any ideas?

**Comment by sshleifer** [comment]
Just added a [commit](https://github.com/huggingface/transfer-learning-conv-ai/pull/29/commits/8477c2e439014e3c59b8f1be6df68b0b08fcffe9) to get evaluation.py working for GPT2 on this branch. It only works in hits@1 mode, the default. Would love to know how it does with your checkpoint!

File: convai_evaluation.py [hunk 2]
```diff
@@ -57,12 +58,14 @@ def __init__(self, opt, shared=None):
             self.logger.info("Get pretrained model and tokenizer")
             if args.model_checkpoint == "":
                 args.model_checkpoint = download_pretrained_model()
-
-            self.tokenizer = OpenAIGPTTokenizer.from_pretrained(args.model_checkpoint)
-            if self.args.eval_type == "hits@1":
-                self.model_checkpoint = OpenAIGPTDoubleHeadsModel.from_pretrained(args.model_checkpoint)
+            if 'gpt2' in args.model_checkpoint:
```
Comment (kiddyboots216) [inline convai_evaluation.py]: I think if someone is passing in a path to a model checkpoint, they're not going to have 'gpt2' in that path name by default. Maybe we can separate model_checkpoint into model_name and model_checkpoint, where model_name can be openai-gpt and gpt2 and model_checkpoint will be "" or a path to a run.

Comment (sshleifer) [reply convai_evaluation.py]: I'm with you 100%, but `train.py` makes the same assumption and I want to keep the PR small. At the very least, `train.py` should generate model checkpoints with 'gpt2' in the path if the next step expects that.

**Comment by kiddyboots216** [comment]
I get `{'exs': 7801, 'hits@1': 0.785, 'hits@5': 0.968, 'hits@10': 0.995, 'hits@100': 1.0}` after finetuning GPT2 for 1 epoch with SGD (lr=0.04, momentum=0.9), so looks like the fix works. Thanks!

File: convai_evaluation.py (hunk 2 above)
Comment (kiddyboots216) [reply convai_evaluation.py]: Ok. We can add a line after line 249 in `train.py` to set `log_dir = os.path.join(log_dir, args.model_checkpoint' perhaps.

File: requirements.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 torch
 pytorch-ignite
-pytorch-pretrained-bert >= 0.6.2
-tensorboardX
-tensorflow  # for tensorboardX
\ No newline at end of file
+pytorch-transformers>=1.2
+tensorboardX==1.8
+tensorflow  # for tensorboardX
```
Comment (julien-c) [inline requirements.txt]: maybe require `torch>=1.2` so you can use the built-in Tensorboard logger?

Comment (kiddyboots216) [reply requirements.txt]: yeah I've been using torch.utils.tensorboard.SummaryWriter locally. you do still need tensorboard as a dependency, though.

Comment (sshleifer) [reply requirements.txt]: ignite v0.2.1 will use the builtin if tensorboardx is not installed (see https://github.com/pytorch/ignite/pull/617). Once that version is released we can delete the tensorboardx requirement.

**Comment by julien-c** [comment]
Thanks, Sam! I squashed your commits but kept the author/committer attribution.

I've also uploaded the new S3 tarball, and I'll update the URL in the next commit.

File: interact.py [hunk 4]
```diff
@@ -60,14 +61,13 @@ def sample_sequence(personality, history, tokenizer, model, args, current_output
...
-        instance, sequence = build_input_from_segments(personality, history, current_output, tokenizer, with_eos=False)
+        instance, _ = build_input_from_segments(personality, history, current_output, tokenizer, with_eos=False)
 
         input_ids = torch.tensor(instance["input_ids"], device=args.device).unsqueeze(0)
         token_type_ids = torch.tensor(instance["token_type_ids"], device=args.device).unsqueeze(0)
 
         logits = model(input_ids, token_type_ids=token_type_ids)
-
-        if "gpt2" == args.model:
+        if isinstance(logits, tuple):  # for gpt2 and maybe others
```
Comment (g-karthik) [inline interact.py]: @sshleifer what are the other values for `args.model` that work here apart from `openai-gpt` and `gpt2`? Are you planning on making changes to `train.py` and `interact.py` to support other models as well?

File: interact.py [hunk 5]
```diff
@@ -76,6 +76,9 @@ def sample_sequence(personality, history, tokenizer, model, args, current_output
         prev = torch.topk(probs, 1)[1] if args.no_sample else torch.multinomial(probs, 1)
         if i < args.min_length and prev.item() in special_tokens_ids:
             while prev.item() in special_tokens_ids:
+                if probs.max().item() == 1:
```
Comment (g-karthik) [inline interact.py]: Isn't this logic assuming that you obtained `prev` by performing argmax?

If `args.no_sample` were `False`, your `prev` wouldn't necessarily be the one with the highest probability. So the warning you print here will be a false one in that scenario, right?

File: train.py [hunk 6]
```diff
@@ -16,12 +16,14 @@
 from ignite.metrics import Accuracy, Loss, MetricsLambda, RunningAverage
 from ignite.contrib.handlers import ProgressBar, PiecewiseLinear
 from ignite.contrib.handlers.tensorboard_logger import TensorboardLogger, OutputHandler, OptimizerParamsHandler
-from pytorch_pretrained_bert import (OpenAIAdam, OpenAIGPTDoubleHeadsModel, OpenAIGPTTokenizer,
-                                     GPT2DoubleHeadsModel, GPT2Tokenizer, WEIGHTS_NAME, CONFIG_NAME)
+from pytorch_transformers import (AdamW, OpenAIGPTDoubleHeadsModel, OpenAIGPTTokenizer,
```
Comment (g-karthik) [inline train.py]: General question: do you plan on migrating this to `transformers` now?

File: train.py [hunk 7]
```diff
@@ -226,24 +245,26 @@ def inference(engine, batch):
...
-        trainer.add_event_handler(Events.EPOCH_COMPLETED, checkpoint_handler, {'mymodel': getattr(model, 'module', model)})  # "getattr" take care of distributed encapsulation
+        checkpoint_handler = ModelCheckpoint(log_dir, 'checkpoint', save_interval=1, n_saved=3)
+        trainer.add_event_handler(Events.EPOCH_COMPLETED, checkpoint_handler, {'mymodel': getattr(model, 'module', model)})  # "getattr" takes care of distributed encapsulation
 
-        torch.save(args, tb_logger.writer.log_dir + '/model_training_args.bin')
-        getattr(model, 'module', model).config.to_json_file(os.path.join(tb_logger.writer.log_dir, CONFIG_NAME))
-        tokenizer.save_vocabulary(tb_logger.writer.log_dir)
+        torch.save(args, log_dir + '/model_training_args.bin')
+        getattr(model, 'module', model).config.to_json_file(os.path.join(log_dir, CONFIG_NAME))
+        tokenizer.save_pretrained(log_dir)
```
Comment (g-karthik) [inline train.py]: Is there a behavioral difference between `save_vocabulary()` and `save_pretrained()`?

File: interact.py (hunk 4 above)
Comment (sshleifer) [reply interact.py]: only those work at the moment, feel free to send a PR adding more support!

File: interact.py (hunk 5 above)
Comment (sshleifer) [reply interact.py]: if there is a probability of 1 on a token, `prev` will always be that token, regardless of `args.no_sample`.

File: train.py (hunk 6 above)
Comment (sshleifer) [reply train.py]: I don't plan on it at the moment, but feel free to send a PR.

File: train.py (hunk 7 above)
Comment (sshleifer) [reply train.py]: From the [docs](https://huggingface.co/transformers/main_classes/tokenizer.html#transformers.PreTrainedTokenizer.save_pretrained),  

> `save_pretrained` makes sure the full tokenizer can then be re-loaded using the from_pretrained() class method.
> Please use `save_pretrained()` to save the full Tokenizer state if you want to reload it using the from_pretrained() class method.

**Comment by g-karthik** [comment]
Thanks for addressing the comments! Could you elaborate on what you're referring to when you say `interact.py` was broken for GPT-2 even before your pull request?
